File: A2Z_striver/1 Basics/1.5 Hashing/highest_occuring_ele_in_arr/highest_occuring_ele_in_arr.cpp
#include <iostream>
#include <vector>
#include <algorithm> // for sort
#include <unordered_map>



using namespace std;



// Header Comment Block

// Problem: Find the most frequent element (Smallest on tie).

// Approaches: Brute Force, Sorting, Hashing.



class Solution {

public:

    // ~ Approach 1: Brute Force (Nested Loops)

    // Analysis: Time O(N^2) | Space O(1)

    int mostFrequentBrute(vector<int>& arr) {

        int n = arr.size();

        int maxFreq = 0;

        int result = -1;



        for (int i = 0; i < n; i++) {

            int currentFreq = 0;

            

            // - Step 1: Count occurrences of arr[i]

            for (int j = 0; j < n; j++) {

                if (arr[j] == arr[i]) {

                    currentFreq++;

                }

            }



            // - Step 2: Update Max

            // - Logic: If freq is greater, update. If equal, pick smaller value.

            if (currentFreq > maxFreq) {

                maxFreq = currentFreq;

                result = arr[i];

            } else if (currentFreq == maxFreq) {

                result = min(result, arr[i]);

            }

        }

        return result;

    }



    // ~ Approach 2: Sorting (Linear Scan)

    // Analysis: Time O(N log N) | Space O(1) (ignoring recursion stack)

    // Logic: Once sorted (10, 10, 20, 20, 30), duplicates are neighbors.

    int mostFrequentSorting(vector<int>& arr) {

        if (arr.empty()) return -1;

        

        // - Step 1: Sort the array

        sort(arr.begin(), arr.end());



        int maxFreq = 0;

        int result = arr[0];

        

        int currentFreq = 1;

        

        // - Step 2: Single Pass

        for (int i = 1; i < arr.size(); i++) {

            if (arr[i] == arr[i - 1]) {

                currentFreq++;

            } else {

                // Sequence broke. Reset.

                currentFreq = 1;

            }



            // - Step 3: Check against max

            // - Logic: Since array is sorted, we process smaller numbers first.

            //   So if we find a tie, we don't update (keeping the smaller/earlier one).

            if (currentFreq > maxFreq) {

                maxFreq = currentFreq;

                result = arr[i];

            }

        }

        return result;

    }



    // ~ Approach 3: Hashing (Optimal)

    // Analysis: Time O(N) | Space O(N)

    int mostFrequentHashing(vector<int>& arr) {

        unordered_map<int, int> freqMap;

        

        // - Step 1: Populate Map

        for (int x : arr) freqMap[x]++;



        int maxFreq = 0;

        int result = -1;



        // - Step 2: Iterate Map

        for (auto it : freqMap) {

            int num = it.first;

            int count = it.second;



            if (count > maxFreq) {

                maxFreq = count;

                result = num;

            } 

            // - Logic: Explicit tie-breaker for unordered data

            else if (count == maxFreq) {

                result = min(result, num);

            }

        }

        return result;

    }

};



int main() {

    ios_base::sync_with_stdio(false);

    cin.tie(NULL);



    Solution sol;

    vector<int> arr = {30, 10, 20, 20, 30, 10, 20, 10}; 

    // 10 appears 3x, 20 appears 3x. 10 is smaller. Result should be 10.



    cout << "1. Brute Force: " << sol.mostFrequentBrute(arr) << endl;

    

    // Note: Sorting modifies the array order

    cout << "2. Sorting:     " << sol.mostFrequentSorting(arr) << endl;

    

    cout << "3. Hashing:     " << sol.mostFrequentHashing(arr) << endl;



    return 0;

}

File: A2Z_striver/1 Basics/1.5 Hashing/highest_occuring_ele_in_arr/highest_occuring_ele_in_arr.py
import sys
from collections import Counter

# Header Comment Block
# Problem: Find the most frequent element (Smallest on tie).
# Approaches: Brute Force (Yours), Sorting, Hashing.

class Solution:

    # ~ Approach 1: Brute Force with Visited Array (Your Code)
    # Analysis: Time O(N^2) | Space O(N)
    def mostFrequentBrute(self, nums):
        n = len(nums)
        maxFr = 0
        maxEle = -1  # Init with infinity for proper min comparison
        visited = [False] * n
        for i in range(n):
            # - Logic: Skip if we already counted this number
            if visited[i]:
                continue

            fr = 0
            for j in range(i, n):
                if nums[i] == nums[j]:
                    fr += 1
                    visited[j] = True

            # - Logic: Update Max (Strict Greater)
            if fr > maxFr:
                maxFr = fr
                maxEle = nums[i]
            # - Logic: Tie-Breaker (Equal Freq) -> Pick Smaller
            elif fr == maxFr:
                maxEle = min(maxEle, nums[i])

        return maxEle

    # ~ Approach 2: Sorting
    # Analysis: Time O(N log N) | Space O(1) (depends on sort impl)
    def mostFrequentSorting(self, nums):
        if not nums:
            return -1

        nums.sort()  # O(N log N)
        n = len(nums)
        maxFr = 0
        maxEle = nums[0]
        currFr = 1

        # - Step: Iterate starting from index 1
        for i in range(1, n):
            if nums[i] == nums[i - 1]:
                currFr += 1
            else:
                currFr = 1  # Reset counter

            # - Logic: Update only if strictly greater.
            # Since array is sorted, we encounter smaller numbers first.
            # So for ties, we automatically keep the smaller one by doing nothing.
            if currFr > maxFr:
                maxFr = currFr
                maxEle = nums[i]

        # Edge case: If array length is 1 or all unique, initial maxEle is correct
        return maxEle if n > 1 else nums[0]

    # ~ Approach 3: Hashing (Optimal)
    # Analysis: Time O(N) | Space O(N)
    def mostFrequentHashing(self, nums):
        # - Step 1: Count Frequencies (One Liner)
        count = Counter(nums)
        maxFr = 0
        maxEle = -1
        # - Step 2: Iterate and Compare
        for num, fr in count.items():
            if fr > maxFr:
                maxFr = fr
                maxEle = num
            elif fr == maxFr:
                maxEle = min(maxEle, num)
        return maxEle


# --- Driver Code ---

if __name__ == "__main__":
    sol = Solution()

    # Test Case: 10 appears 2x, 20 appears 2x. Winner should be 10.
    arr = [20, 10, 20, 10, 30]

    print(f"1. Brute Force: {sol.mostFrequentBrute(arr)}")

    print(f"2. Sorting:     {sol.mostFrequentSorting(list(arr))}")  # Pass copy to avoid sorting original

    print(f"3. Hashing:     {sol.mostFrequentHashing(arr)}")
